### PR TITLE
F/aragondao handler

### PR DESCRIPTION
**Pull request by nigeon**
Added AragonDAO's handler for the faucet.
Works with multisig, token and addresslist daos.
Multi-network.

### CODE REVIEW

File: aragondaohandler/handler.go [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+		currentDate.Month() != parsedDate.Month() ||
+		currentDate.Day() != parsedDate.Day() {
+		return addr, errors.New("signature date is not from today")
+	}
+
+	return addr, nil
+}
+
+func IsAragonDaoAddress(addr common.Address, network string) (bool, error) {
+	if network != "mainnet" && network != "goerli" && network != "sepolia" && network != "polygon" && network != "mumbai" {
```
Comment (p4u) [inline aragondaohandler/handler.go:48]: Oh! please make this list of networks a global variable (a map[string]struct{} works and its efficient)

File: aragondaohandler/handler.go [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+			addresslistVotingPlugins(where: {members_: {address: $address}}) {
+				members(where: {address: $address}) {
+				address
+				}
+			}
+		}
+	`)
+
+	// use the network to determine the subgraphURL
+	graphURL := "https://subgraph.satsuma-prod.com/qHR2wGfc5RLi6/aragon/osx-" + string(network) + "/version/v1.3.0/api"
```
Comment (p4u) [inline aragondaohandler/handler.go:73]: This should be a constant, you an use a custom tag within the string such as {NETWORK} and replace it here.

File: aragondaohandler/handler.go [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+	}
+	defer resp.Body.Close()
+
+	var body []byte
+	if body, err = ioutil.ReadAll(resp.Body); err != nil {
+		return false, err
+	}
+
+	// Parse the JSON response
+	type Response struct {
```
Comment (p4u) [inline aragondaohandler/handler.go:109]: Define the types in a file "types.go" part of the aragonhandler package

Comment (p4u) [reply aragondaohandler/handler.go:109]: You can also add the network map and the URL constant within the same file.

File: handlers.go [hunk 4]
```diff
@@ -158,9 +167,70 @@ func (f *faucet) authOAuthUrl(msg *apirest.APIdata, ctx *httprouter.HTTPContext)
...
+	}
+
+	// Check if the address is already funded
+	if funded, t := f.storage.checkIsFundedAddress(addr); funded {
+		errReason := fmt.Sprintf("address %s already funded, wait until %s", addr.Hex(), t)
+		return ctx.Send(new(HandlerResponse).SetError(errReason).MustMarshall(), CodeErrFlood)
+	}
+
+	// Check if the address is an Aragon DAO address by checking to AragonGraphQL
+	if newRequest.Network != "" {
```
Comment (p4u) [inline handlers.go:208]: This is a bit weird, I'd either force the user to define network or not (and check all configured networks until one is found). Probably the second is easier for the UI side.

**Review by p4u — CHANGES_REQUESTED** [review]

File: handlers.go (hunk 4 above)
Comment (nigeon) [reply handlers.go:208]: The reason behind this is offering both possibilities,
1. Searching in just one specific network
2. Searching in any network

In the scaffold, we don't know (yet) the network that should be used (as it's "network-agnostic"). I think this is not the case for the Aragon Plugin.

Anyway, will ask about the logic to @jpaulet

**Review by p4u — APPROVED** [review]

**Pull request merged**